### PR TITLE
Orphan Pool

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- Added an implementation of the OrphanPool
- Added tests for the OrphanPool
- Added contains function to MerkleMountainRange
- Cleaned TxPriority and added get_fee_per_byte function to transaction 
- Modified the create_test_tx, create_test_block, create_test_kernel functions so that they can be used for the OrphanPool Tests
- Added create_test_utxos and extend_test_utxos test_utils functions

## Motivation and Context
The Orphan Pool contains all the received transactions that attempt to spend UTXOs that don't exist.

## How Has This Been Tested?
New tests have been added.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/transaction.rs [hunk 1]
```diff
@@ -503,9 +503,15 @@ impl Transaction {
         &self.body
     }
 
+    /// Returns the byte size or weight of a transaction
     pub fn get_weight(&self) -> Result<usize, MessageFormatError> {
         Ok(self.to_binary()?.len())
     }
+
+    /// Returns the total fee allocated to each byte of the transaction
+    pub fn get_fee_per_byte(&self) -> Result<f64, MessageFormatError> {
```
Comment (CjS77) [inline base_layer/core/src/transaction.rs]: this is called `fee_per_gram` in other parts of the codebase.

File: base_layer/core/src/transaction.rs [hunk 2]
```diff
@@ -503,9 +503,15 @@ impl Transaction {
         &self.body
     }
 
+    /// Returns the byte size or weight of a transaction
     pub fn get_weight(&self) -> Result<usize, MessageFormatError> {
```
Comment (CjS77) [inline base_layer/core/src/transaction.rs]: This isn't quite right.

The weight calculation is actually in [fee.rs#L35], though it's still bundles in with the fee calculation.

We should just extract that out; have
```
pub fn calculate_weight(num_inputs: usize, num_outputs: usize) -> u64 {
          COST_PER_INPUT * num_inputs as u64 + COST_PER_OUTPUT * num_outputs as u64
    }
```

File: base_layer/core/src/transaction.rs (hunk 1 above)
Comment (CjS77) [reply base_layer/core/src/transaction.rs]: This isn't 100% summetrical with the `calculate_fee` function, but that's ok, if we just name this `calculate_ave_fee_per_gram` or something,

File: base_layer/core/src/mempool/orphan_pool/orphan_pool.rs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    pub fn has_tx_with_excess_sig(&self, excess_sig: &Signature) -> bool {
+        self.txs_by_signature.contains_key(excess_sig)
+    }
+
+    /// Check if the required UTXOs have been created and if the status of any of the transactions in the OrphanPool has
+    /// changed. Remove valid transactions and valid transactions with time-locks from the OrphanPool.
+    // TODO: A reference to the UTXO set should not be passed in like this, but rather a handle or stream to the Chain
+    // should be provided to the OrphanPool during creation allowing a Chain call to be used to query the current
+    // block_height and UTXO set
+    pub fn remove_valid(
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/mempool/orphan_pool/orphan_pool.rs]: I agree, change applied.

File: base_layer/core/src/transaction.rs (hunk 1 above)
Comment (neonknight64) [reply base_layer/core/src/transaction.rs]: Function renamed.

File: base_layer/core/src/transaction.rs (hunk 2 above)
Comment (neonknight64) [reply base_layer/core/src/transaction.rs]: I wondered if there wasn't a formula. Change applied, but I hope I interpreted your suggestion correctly.

File: base_layer/core/src/fee.rs [hunk 4]
```diff
@@ -31,9 +31,7 @@ pub const BASE_COST: u64 = 1;
 impl Fee {
     /// Computes the absolute transaction fee given the fee-per-gram, and the size of the transaction
     pub fn calculate(fee_per_gram: MicroTari, num_inputs: usize, num_outputs: usize) -> MicroTari {
-        (BASE_COST +
-            (COST_PER_INPUT * num_inputs as u64 + COST_PER_OUTPUT * num_outputs as u64) * u64::from(fee_per_gram))
-        .into()
+        (BASE_COST + Fee::calculate_weight(num_inputs, num_outputs) * u64::from(fee_per_gram)).into()
```
Comment (CjS77) [inline base_layer/core/src/fee.rs]: I think you can multiply a `MicroTari` by a u64.. maybe the order of operands matters

File: base_layer/core/src/fee.rs [hunk 5]
```diff
@@ -46,4 +44,9 @@ impl Fee {
             fee
         }
     }
+
+    /// Calculate the weight of a transaction based on the number of inputs and outputs
+    pub fn calculate_weight(num_inputs: usize, num_outputs: usize) -> u64 {
+        COST_PER_INPUT * num_inputs as u64 + COST_PER_OUTPUT * num_outputs as u64
+    }
```
Comment (CjS77) [inline base_layer/core/src/fee.rs]: Nice. WEIGHT_PER_xxxPUT is a better name now, I guess.

File: base_layer/core/src/transaction.rs [hunk 6]
```diff
@@ -503,8 +500,14 @@ impl Transaction {
...
 
-    pub fn get_weight(&self) -> Result<usize, MessageFormatError> {
-        Ok(self.to_binary()?.len())
+    /// Returns the byte size or weight of a transaction
+    pub fn calculate_weight(&self) -> u64 {
+        Fee::calculate_weight(self.body.inputs.len(), self.body.outputs.len())
+    }
+
+    /// Returns the total fee allocated to each byte of the transaction
+    pub fn calculate_ave_fee_per_gram(&self) -> Result<f64, MessageFormatError> {
```
Comment (CjS77) [inline base_layer/core/src/transaction.rs]: How can this return an Error?

Shouldn't it just return f64?

File: base_layer/core/src/fee.rs (hunk 4 above)
Comment (neonknight64) [reply base_layer/core/src/fee.rs]: I shuffled things around, but it didn't seem to work without the explicit conversion.

File: base_layer/core/src/fee.rs (hunk 5 above)
Comment (neonknight64) [reply base_layer/core/src/fee.rs]: Function name changed to `weight_per_xxxput`, it still feels like there should be a better name.

File: base_layer/core/src/transaction.rs (hunk 6 above)
Comment (neonknight64) [reply base_layer/core/src/transaction.rs]: Good catch, change applied

File: base_layer/core/src/fee.rs (hunk 5 above)
Comment (CjS77) [reply base_layer/core/src/fee.rs]: Lol. No, I meant

pub fn calculate_weight(num_inputs: usize, num_outputs: usize) -> u64 {
        WEIGHT_PER_INPUT * num_inputs as u64 + WEIGHT_PER_OUTPUT * num_outputs as u64
    }

Comment (CjS77) [reply base_layer/core/src/fee.rs]: The original name was fine

Comment (neonknight64) [reply base_layer/core/src/fee.rs]: Oops, change applied correctly.

**Pull request merged**